kernel/src/hil/public_key_crypto/keys.rs

```diff
@@ -0,0 +1,90 @@
+    fn generation_complete(&'a self, result: Result<(), ErrorCode>);
+}
+
+/// An internal representation of a asymetric Public key.
+///
+/// This trait is useful for managing keys internally in Tock.
+pub trait PubKey<'a> {
+    /// Import an existing public key. This will override any existing keys.
+    ///
+    /// The possible ErrorCodes are:
+    ///     - `ALREADY`: A key is already imported or in the process of being
+    ///                  generated.
+    ///     - `INVAL`: An invalid key was supplied.
+    ///     - `NOMEM`: The key can not be stored.
+    fn import_public_key(&mut self, public_key: &[u8]) -> Result<(), ErrorCode>;
+
+    /// Store the public key in buffer if it exists.
+    fn pub_key(&'a self, buffer: &mut [u8]) -> Option<()>;
+
+    /// Report the length of the public key, as returned from `pub_key()`.
+    fn len(&self) -> usize;
+}
+
+/// An internal representation of a asymetric Public and Private key.
+///
+/// This trait is useful for managing keys internally in Tock.
+pub trait PubPrivKey<'a>: PubKey<'a> {
+    /// Import an existing private key. This will override any existing keys.
+    ///
+    /// The possible ErrorCodes are:
+    ///     - `ALREADY`: A key is already imported or in the process of being
+    ///                  generated.
+    ///     - `INVAL`: An invalid key was supplied.
+    ///     - `NOMEM`: The key can not be stored.
+    fn import_private_key(&mut self, private_key: &[u8]) -> Result<(), ErrorCode>;
+
+    /// Store the private key in buffer if it exists.
+    fn priv_key(&'a self, buffer: &mut [u8]) -> Option<()>;
+
+    /// Report the length of the private key, as returned from `priv_key()`.
+    fn len(&self) -> usize;
+}
+
+/// An internal representation of generating asymetric Public/Private key
+/// pairs.
+///
+/// This trait is useful for managing keys internally in Tock.
+pub trait PubPrivKeyGenerate<'a>: PubPrivKey<'a> {
+    /// Set the client. This client will be called when the `generate()`
+    /// function is complete. If using an existing key this doesn't need to be
+    /// used.
+    fn set_client(&'a self, client: &'a dyn PubPrivKeyGenerateClient<'a>);
+
+    /// This generates a new private/public key pair of `length` bits.
+    /// This will call the `generation_complete()` on completion. They keys
+    /// cannot be used and will return `None` until the upcall has been called.
+    ///
+    /// The keys generated by `generate()` will depend on the implementation.
+    ///
+    /// The possible ErrorCodes are:
+    ///     - `BUSY`: The underlying `trng` is busy.
+    ///     - `ALREADY`: A key is already imported or in the process of being
+    ///                  generated.
+    ///     - `OFF`: The underlying `trng` is powered down.
+    ///     - `INVAL`: An invalid length was supplied
+    ///     - `NOMEM`: The key can not be stored.
+    fn generate(&'a self, trng: &'a dyn entropy::Entropy32, length: usize)
```

Comment:
Where is it stored? How does one know the right/maximum length? E.g., suppose there is no key currently. Can you call `priv_key` to get the maximum length, or will it return 0 because there is no key? There should be a `max_length` method.

Reply:
OK, looking at the RSA implementation, it seems like there is a constant key length, which defines the only size of the key allowed. So a key structure is bound to a specific key/length. So why can generate take a length? The traits need to describe this so that the meaning of the methods is clear. If the traits are intended to support both variable and static length keys, then they need some more support for the variable case.